doc/code/curves.md

```diff
@@ -0,0 +1,217 @@
+
+
+## Curve Types
+
+This section provides an overview over the available curves types.
+
+### Primitive
+
+![Primitive curve types UML](images/primitive_curves_uml.svg)
+
+Primitive curves are intended for members with single value types. These include, for example,
+the primitive C++ types (e.g. `int`, `float`, `std::string`), simple structs and data classes,
+and shared pointers.
+
+On contruction of a primitive curve object, a keyframe with time `t = time::time_t::min_value()`
+and the default value of the value type is inserted. This is done to ensure that for any
+requested time `t`, there is always a valid value to be returned. This mirrors the expected
+behaviour from declaring primitive values in C++ where members may be auto-initialized without
+explicit assignment to a default value. The default value for curves can also be explicitely
+assigned in the constructor. For value types that don't have default values, a default value
+must always be passed in the constructor.
+
+`BaseCurve` objects can be targeted by or trigger events from the [event system](/doc/code/event_system.md).
+As a consequence, curves are not copy constructable as they require a unique ID for
+event management. However, it is possible to copy the keyframes of one curve to
+the other using the `sync(..)` method. `sync(..)` also works for curves with different
+values types if a converter function from one value type to the other is supplied.
+
+#### Common Operations
+
+All primitive curves support the following operations. They may work slightly different
+for specific curve types.
+
+**Read**
+
+| Method          | Description                                                 |
+| --------------- | ----------------------------------------------------------- |
+| `get(t)`        | Get (interpolated) value at time `t`                        |
+| `frame(t)`      | Get the previous keyframe (time and value) before or at `t` |
+| `next_frame(t)` | Get the next keyframe (time and value) after `t`            |
+
+**Modify**
+
+| Method                  | Description                                                                       |
+| ----------------------- | --------------------------------------------------------------------------------- |
+| `set_insert(t, value)`  | Insert a new keyframe value at time `t`                                           |
+| `set_last(t, value)`    | Insert a new keyframe value at time `t`; delete all keyframes after time `t`      |
+| `set_replace(t, value)` | Insert a new keyframe value at time `t`; remove all other keyframes with time `t` |
+
+**Copy**
+
+| Method           | Description                                                                            |
+| ---------------- | -------------------------------------------------------------------------------------- |
+| `sync(Curve, t)` | copy keyframes from `Curve`, starting at time `t`; delete all keyframes after time `t` |
```

Comment:
delete from where? is the target curve cleared as well?

Reply:
should be fixed and more understandable now